core/src/apps/cardano/sign_tx/signer.py

```diff
@@ -461,23 +446,25 @@ async def _process_legacy_output(
     async def _process_babbage_output(
         self,
         output_dict: HashBuilderDict[int, Any],
-        output: messages.CardanoTxOutput,
+        output: CardanoTxOutput,
         should_show: bool,
     ) -> None:
         """
         This output format corresponds to the post-Alonzo format in CDDL.
         Note that it is to be used also for outputs with no Plutus elements.
         """
+        from ..helpers.hash_builder_collection import HashBuilderEmbeddedCBOR
+
         address = self._get_output_address(output)
-        output_dict.add(BABBAGE_OUTPUT_KEY_ADDRESS, address)
+        output_dict.add(_BABBAGE_OUTPUT_KEY_ADDRESS, address)
```

Comment:
the `with` statement does things with the return value, so caching the attribute should be exactly as fine as in other usages

Reply:
And what is the purpose here when we do not seem to consume the return value like we would in `with add(...) as return_value`?

Reply:
the `return_value` in this case is actually the second argument, so e.g. `output_value_list`. we don't need to grab it `as anything` because we already have it

as for the purpose of all this, I forget the details but it's something to do with building a hash of a CBOR-encoded structure on the fly. the `with` statement is telling the hasher that we are now diving into a sub-element of the parent dict, and there is some sort of check to ensure that we inserted (and hashed) all the expected values before continuing to add things to the parent dict

Reply:
Cached in [5431b25](https://github.com/trezor/trezor-firmware/pull/2525/commits/5431b256f3f00062917631501b9033bf90f07e17)